spring-core/src/main/java/org/springframework/core/retry/support/CompositeRetryListener.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2002-2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.core.retry.support;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import org.springframework.core.retry.RetryExecution;
+import org.springframework.core.retry.RetryListener;
+import org.springframework.core.retry.RetryTemplate;
+import org.springframework.util.Assert;
+
+/**
+ * A composite implementation of the {@link RetryListener} interface. This class
+ * is used to compose multiple listeners within a {@link RetryTemplate}.
+ *
+ * <p>Delegate listeners will be called in their registration order.
+ *
+ * @author Mahmoud Ben Hassine
+ * @since 7.0
+ */
+public class CompositeRetryListener implements RetryListener {
+
+	private final List<RetryListener> listeners;
+
+	/**
+	 * Create a new {@link CompositeRetryListener}.
+	 */
+	public CompositeRetryListener() {
+		this.listeners = new LinkedList<>();
+	}
+
+	/**
+	 * Create a new {@link CompositeRetryListener} with a list of delegates.
+	 * @param listeners the delegate listeners to register
+	 */
+	public CompositeRetryListener(List<RetryListener> listeners) {
+		this.listeners = listeners;
```

Comment:
This is not correct.
We have to create a copy here internally.
Otherwise, any mutate to that external collection would effect our component as well.
And there is chance of opposite, more dangerous logic: we modify `this.listeners` internally here and everything else using that external collection would be effected.
I remember having this error a lot in our projects when we used SonarQube.